File: surf-friends/src/main/java/dev/slne/surf/friends/FriendManager.java
package dev.slne.surf.friends;

import com.github.benmanes.caffeine.cache.Caffeine;
import com.github.benmanes.caffeine.cache.LoadingCache;

import dev.slne.surf.friends.database.Database;
import dev.slne.surf.friends.util.PluginColor;

import it.unimi.dsi.fastutil.objects.ObjectArrayList;
import it.unimi.dsi.fastutil.objects.ObjectList;

import java.util.Objects;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;

import lombok.Getter;
import lombok.experimental.Accessors;

import net.kyori.adventure.text.Component;

import org.bukkit.Bukkit;
import org.bukkit.OfflinePlayer;
import org.bukkit.entity.Player;
import org.jetbrains.annotations.NotNull;

@Getter
@Accessors(fluent = true)
public class FriendManager {
  @Getter
  private static final FriendManager instance = new FriendManager();
  private final LoadingCache<UUID, FriendData> cache = Caffeine.newBuilder().build(FriendManager::loadFriendData);

  /* Friend Management */

  public void addFriend(UUID player, UUID target) {
    FriendData playerData = this.queryFriendData(player);
    FriendData targetData = this.queryFriendData(target);

    if(playerData.getFriends().contains(target)) {
      return;
    }

    if(targetData.getFriends().contains(player)) {
      return;
    }

    playerData.getFriends().add(target);
    cache.put(player, playerData);

    targetData.getFriends().add(player);
    cache.put(target, targetData);

    this.sendMessage(player, Component.text("Du bist nun mit ")
        .append(Component.text(this.getName(target), PluginColor.GOLD))
        .append(Component.text(" befreundet.")));
    this.sendMessage(target, Component.text("Du bist nun mit ")
        .append(Component.text(this.getName(player), PluginColor.GOLD))
        .append(Component.text(" befreundet.")));

  }

  public void removeFriend(UUID player, UUID target) {
    FriendData playerData = this.queryFriendData(player);
    FriendData targetData = this.queryFriendData(target);

    if(!playerData.getFriends().contains(target)) {
      return;
    }

    if(!targetData.getFriends().contains(player)) {
      return;
    }


    playerData.getFriends().remove(target);
    cache.put(player, playerData);

    targetData.getFriends().remove(player);
    cache.put(target, targetData);

    this.sendMessage(player, Component.text("Du hast ")
        .append(Component.text(this.getName(target), PluginColor.GOLD))
        .append(Component.text(" als Freund entfernt.")));
    this.sendMessage(target, Component.text("Du wurdest von ")
        .append(Component.text(this.getName(player), PluginColor.GOLD))
        .append(Component.text(" als Freund entfernt.")));

  }

  public void sendFriendRequest(UUID player, UUID target) {
    FriendData targetData = this.queryFriendData(target);

    if (targetData.getFriendRequests().contains(player)) {
      return;
    }

    targetData.getFriendRequests().add(player);
    cache.put(target, targetData);

    this.sendMessage(player, Component.text("Du hast eine Freundschaftsanfrage an ")
        .append(Component.text(this.getName(target), PluginColor.GOLD))
        .append(Component.text(" gesendet.")));

    if(targetData.getAllowRequests()){
      this.sendMessage(target, Component.text("Du hast eine Freundschaftsanfrage von ")
          .append(Component.text(this.getName(player), PluginColor.GOLD))
          .append(Component.text(" erhalten.")));
    }
  }

  public void acceptFriendRequests(UUID player, UUID target) {
    FriendData playerData = this.queryFriendData(player);
    if (!playerData.getFriendRequests().contains(target)) {
      return;
    }

    playerData.getFriendRequests().remove(target);
    cache.put(player, playerData);

    this.addFriend(player, target);

    this.sendMessage(player, Component.text("Du hast die Freundschaftsanfrage von ")
        .append(Component.text(this.getName(target), PluginColor.GOLD))
        .append(Component.text(" akzeptiert.")));
    this.sendMessage(target, Component.text("Die Freundschaftsanfrage an ")
        .append(Component.text(this.getName(player), PluginColor.GOLD))
        .append(Component.text(" wurde akzeptiert.")));
  }

  public void denyFriendRequest(UUID player, UUID target) {
    FriendData playerData = this.queryFriendData(player);

    playerData.getFriendRequests().remove(target);
    cache.put(player, playerData);

    this.sendMessage(player, Component.text("Du hast die Freundschaftsanfrage von ")
        .append(Component.text(this.getName(target), PluginColor.GOLD))
        .append(Component.text(" abgelehnt.")));
    this.sendMessage(target, Component.text("Die Freundschaftsanfrage an ")
        .append(Component.text(this.getName(player), PluginColor.GOLD))
        .append(Component.text(" wurde abgelehnt.")));
  }

  public boolean areFriends(UUID player, UUID target) {
    return this.queryFriendData(player).getFriends().contains(target);
  }

  public boolean hasFriendRequest(UUID player, UUID target) {
    return this.queryFriendData(player).getFriendRequests().contains(target);
  }

  public ObjectList<UUID> getFriends(UUID player) {
    return this.queryFriendData(player).getFriends();
  }

  public ObjectList<UUID> getFriendRequests(UUID player) {
    return this.queryFriendData(player).getFriendRequests();
  }

  public boolean toggle(UUID player) {
    FriendData playerData = this.queryFriendData(player);

    playerData.setAllowRequests(!playerData.getAllowRequests());
    cache.put(player, playerData);

    return playerData.getAllowRequests();
  }

  public boolean isAllowingRequests(UUID player) {
    return this.queryFriendData(player).getAllowRequests();
  }

  public FriendData queryFriendData(UUID player) {
    if (cache.get(player) == null) {
      return newFriendData(player);
    }

    return cache.get(player);
  }

  public static FriendData loadFriendData(UUID player) {
    Database.getFriendData(player).thenApply(friendData -> {
      if (friendData == null) {
        return newFriendData(player);
      }

      return friendData;
    });
    return null;
  }

  public static @NotNull CompletableFuture<FriendData> loadFriendDataAsync(UUID player) {
    return Database.getFriendData(player).thenApply(friendData -> Objects.requireNonNullElseGet(friendData, () -> newFriendData(player)));
  }


  public CompletableFuture<Void> saveFriendData(UUID player) {
    return CompletableFuture.runAsync(() -> {
      FriendData friendData = this.queryFriendData(player);
      Database.saveFriendData(friendData).join();

      cache.invalidate(player);
    });
  }

  public CompletableFuture<Void> saveAll(boolean closeConnection) {
    ObjectList<CompletableFuture<Void>> futures = new ObjectArrayList<>();

    for (UUID player : cache.asMap().keySet()) {
      CompletableFuture<Void> future = this.saveFriendData(player);
      futures.add(future);
    }

    CompletableFuture<Void> allSaves = CompletableFuture.allOf(futures.toArray(new CompletableFuture[0]));

    if (closeConnection) {
      return allSaves.thenCompose(v -> CompletableFuture.runAsync(Database::closeConnection));
    } else {
      return allSaves;
    }
  }


  public void sendPlayer(UUID player, UUID target) {
    // TODO: Send to Server
  }

  private void sendMessage(UUID uuid, Component message) {
    Player player = Bukkit.getPlayer(uuid);
    if (player != null) {
      player.sendMessage(SurfFriendsPlugin.getPrefix().append(message));
    }
  }
  
  private String getName(UUID uuid) {
    OfflinePlayer offlinePlayer = Bukkit.getOfflinePlayer(uuid);
    
    return offlinePlayer.getName() == null ? "Unbekannt" : offlinePlayer.getName();
  }

  public static FriendData newFriendData(UUID player){
    return new FriendData(player, new ObjectArrayList<>(), new ObjectArrayList<>(), true);
  }

  public ObjectList<Player> getOnlineFriends(UUID player) {
    //TODO: Cloud implementation

    return null;
  }
}

File: surf-friends/src/main/java/dev/slne/surf/friends/FriendManager.kt
@file:OptIn(DelicateCoroutinesApi::class)

package dev.slne.surf.friends

import com.github.benmanes.caffeine.cache.Caffeine
import dev.hsbrysk.caffeine.CoroutineLoadingCache
import dev.hsbrysk.caffeine.buildCoroutine
import dev.slne.surf.friends.database.Database
import dev.slne.surf.friends.listener.util.PluginColor
import it.unimi.dsi.fastutil.objects.ObjectArrayList
import it.unimi.dsi.fastutil.objects.ObjectList
import it.unimi.dsi.fastutil.objects.ObjectSet
import kotlinx.coroutines.DelicateCoroutinesApi
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import net.kyori.adventure.text.Component
import net.kyori.adventure.text.event.ClickEvent
import org.bukkit.Bukkit
import org.bukkit.entity.Player
import java.util.*

object FriendManager {
    val cache: CoroutineLoadingCache<UUID, FriendData> =
        Caffeine.newBuilder().buildCoroutine { player: UUID -> loadFriendData(player) }

    /* Friend Management */
    suspend fun addFriend(player: UUID, target: UUID) {
        val playerData = queryFriendData(player)
        val targetData = queryFriendData(target)

        if(this.areFriends(player, target)) {
            return
        }

        playerData.friends.add(target);
        targetData.friends.add(player)

        cache.put(player, playerData)
        cache.put(target, targetData)

        sendMessage(player, Component.text("Du bist nun mit ")
                .append(Component.text(getName(target), PluginColor.GOLD))
                .append(Component.text(" befreundet."))
        )

        sendMessage(target, Component.text("Du bist nun mit ")
                .append(Component.text(getName(player), PluginColor.GOLD))
                .append(Component.text(" befreundet."))
        )
    }

    suspend fun removeFriend(player: UUID, target: UUID) {
        val playerData = queryFriendData(player)
        val targetData = queryFriendData(target)

        if (!playerData.friends.remove(target)) {
            return
        }

        if (!targetData.friends.remove(player)) {
            return
        }

        cache.put(player, playerData)
        cache.put(target, targetData)

        sendMessage(
            player, Component.text("Du hast ")
                .append(Component.text(getName(target), PluginColor.GOLD))
                .append(Component.text(" aus deiner Freundesliste entfernt."))
        )

        sendMessage(
            target, Component.text("Du wurdest von ")
                .append(Component.text(getName(player), PluginColor.GOLD))
                .append(Component.text("´s Freundesliste entfernt."))
        )
    }

    suspend fun sendFriendRequest(player: UUID, target: UUID) {
        val targetData = queryFriendData(target)

        if (targetData.friends.contains(player)) {
            return
        }

        if(targetData.friendRequests.contains(player)) {
            return
        }

        if(player == target) {
            return
        }

        if (!targetData.friendRequests.add(player)) {
            return
        }

        cache.put(target, targetData)

        sendMessage(player, Component.text("Du hast eine Freundschaftsanfrage an ")
                .append(Component.text(getName(target), PluginColor.GOLD))
                .append(Component.text(" gesendet."))
        )

        if (targetData.allowRequests) {
            sendMessage(target, Component.text("Du hast eine Freundschaftsanfrage von ")
                    .append(Component.text(getName(player), PluginColor.GOLD))
                    .append(Component.text(" erhalten. "))
                    .append(Component.text("[AKZEPTIEREN]", PluginColor.LIGHT_GREEN).clickEvent(ClickEvent.runCommand("/friend accept " + getNameOrUUID(player))))
                    .append(Component.text(" [ABLEHNEN]", PluginColor.RED).clickEvent(ClickEvent.runCommand("/friend deny " + getNameOrUUID(player))))
            )
        }
    }

    suspend fun acceptFriendRequest(player: UUID, target: UUID) {
        val playerData = queryFriendData(player)
        val targetData = queryFriendData(target)

        playerData.friendRequests.remove(target);
        targetData.friendRequests.remove(player)

        cache.put(player, playerData)
        cache.put(target, targetData)

        addFriend(player, target)

        sendMessage(
            player, Component.text("Du hast die Freundschaftsanfrage von ")
                .append(Component.text(getName(target), PluginColor.GOLD))
                .append(Component.text(" akzeptiert."))
        )

        sendMessage(
            target, Component.text("Deine Freundschaftsanfrage an ")
                .append(Component.text(getName(player), PluginColor.GOLD))
                .append(Component.text(" wurde akzeptiert."))
        )
    }

    suspend fun denyFriendRequest(player: UUID, target: UUID) {
        val playerData = queryFriendData(player)
        val targetData = queryFriendData(target)

        playerData.friendRequests.remove(target)
        targetData.friendRequests.remove(player)
        cache.put(player, playerData)
        cache.put(target, targetData)

        sendMessage(
            player, Component.text("Du hast die Freundschaftsanfrage von ")
                .append(Component.text(getName(target), PluginColor.GOLD))
                .append(Component.text(" abgelehnt."))
        )

        sendMessage(
            target, Component.text("Deine Freundschaftsanfrage an ")
                .append(Component.text(getName(player), PluginColor.GOLD))
                .append(Component.text(" wurde abgelehnt."))
        )
    }

    suspend fun hasFriendRequest(player: UUID, target: UUID): Boolean =
        queryFriendData(player).friendRequests.contains(target)

    suspend fun getFriendRequests(player: UUID): ObjectSet<UUID> =
        queryFriendData(player).friendRequests

    suspend fun toggle(player: UUID): Boolean {
        val playerData = queryFriendData(player)

        playerData.allowRequests = !(playerData.allowRequests)
        cache.put(player, playerData)

        return playerData.allowRequests
    }

    suspend fun isAllowingRequests(player: UUID): Boolean = queryFriendData(player).allowRequests

    private suspend fun queryFriendData(player: UUID): FriendData =
        cache.get(player) ?: newFriendData(player)

    suspend fun saveFriendData(player: UUID) = Database.saveFriendData(queryFriendData(player))

    suspend fun saveAll(closeConnection: Boolean = true) {
        cache.synchronous().asMap().map { player -> saveFriendData(player.key) }

        if (closeConnection) {
            withContext(Dispatchers.IO) {
                Database.closeConnection()
            }
        }
    }

    fun sendMessage(uuid: UUID, message: Component = Component.text("???")) {
        val player = Bukkit.getPlayer(uuid) ?: return

        player.sendMessage(prefix.append(message))
    }

    private fun getName(uuid: UUID) = Bukkit.getOfflinePlayer(uuid).name ?: "Unbekannt"
    private fun getNameOrUUID(uuid: UUID) = Bukkit.getOfflinePlayer(uuid).name ?: uuid

    suspend fun getOnlineFriends(player: UUID): ObjectList<Player> {
        // TODO: Cloud implementation
        return ObjectArrayList()
    }

    suspend fun getServer(player: UUID): String {
        // TODO: Cloud implementation
        return "???"
    }

    suspend fun getFriends(player: UUID): ObjectSet<UUID> = queryFriendData(player).friends

    suspend fun areFriends(player: UUID, target: UUID): Boolean =
        queryFriendData(player).friends.contains(target)

    suspend fun loadFriendData(player: UUID): FriendData = Database.getFriendData(player)
    
    fun newFriendData(player: UUID): FriendData = FriendData(player)
}
